Add functions in sanitization to handler ProtocolRPCRequest and ProtocolMessage - Closes#1017

ishantiw opened the pull request:
### Description

Add sanitization functions to handle protocol RPC requests and messages

### Review checklist

* The PR resolves #1017 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

ishantiw committed:
:recycle: Create sanitization functions and use it in peer handle protocol rpc and message

shuse2 commented on packages/lisk-p2p/src/sanitization.ts:
```diff
@@ -92,3 +97,32 @@ export const sanitizePeerInfoList = (
 	}
 };
+
+export const sanitizeRPCRequest = (request: unknown): ProtocolRPCRequest => {
```
instead of `sanitize`, can we use `validate`?

shuse2 commented on packages/lisk-p2p/src/errors.ts:
```diff
@@ -69,3 +69,17 @@ export class RequestFailError extends VError {
 	}
 }
+
+export class InvalidRPCRequest extends VError {
```
prefer `InvalidRPCRequestError`

shuse2 commented on packages/lisk-p2p/src/errors.ts:
```diff
@@ -69,3 +69,17 @@ export class RequestFailError extends VError {
+	}
+}
+
+export class InvalidProtocolMessage extends VError {
```
prefer `InvalidProtocolMessageError`

ishantiw committed:
:nail_care: Change custom error names and rpc function name

ishantiw committed:
:arrow_up: Upgrade socketcluster-server to 14.2.1

ishantiw committed:
:house: Update imports related to sanitization

ishantiw commented:
Addressed all the comments and rebased. Also, I have included the changes related to `socketcluster-server` library in #1021, where I forgot to upgrade the version of the library

ishantiw committed:
:house: Resolve conflicts of interface name changes

jondubois commented on packages/lisk-p2p/src/peer.ts:
```diff
@@ -106,49 +104,34 @@ export class Peer extends EventEmitter {
+				);
 
-				return;
+				this.emit(EVENT_REQUEST_RECEIVED, request);
```
It's dangerous to emit within a `try` block because event listeners which are declared outside this class will be called synchronously; so if a listener is attached to the Peer from outside; for example using `peer.on(EVENT_REQUEST_RECEIVED, myListenerFunction)`; it means that if `myListenerFunction` throws an error or has a bug then it will cause the catch block below to execute and `this.emit(EVENT_INVALID_REQUEST_RECEIVED, packet);` will be called even though the request is fine; makes it very hard to debug.

An alternative approach might be to `return` from inside the `catch` block just after calling `this.emit(EVENT_INVALID_REQUEST_RECEIVED, packet);` and on the next line outside the catch block (which will only run if request is valid), we can emit the request with `this.emit(EVENT_REQUEST_RECEIVED, request);`.

It's easy to do with `let` but may be kind of tricky to do with only `const`. Maybe it's worth separating the validation logic from the sanitization logic and only have the try-catch block around the validation part?

Another option would be to emit asynchronously (e.g. inside a setTimeout/setImmediate) but that is more ugly and computationally expensive.

jondubois commented on packages/lisk-p2p/src/peer.ts:
```diff
@@ -106,49 +104,34 @@ export class Peer extends EventEmitter {
-				return;
+			try {
+				const protocolMessage = sanitizeProtocolMessage(packet);
+				this.emit(EVENT_MESSAGE_RECEIVED, protocolMessage);
```
Same problem as above.

jondubois commented on packages/lisk-p2p/src/sanitization.ts:
```diff
@@ -92,3 +98,32 @@ export const sanitizePeerInfoList = (
+		throw new InvalidRPCRequestError('Invalid request');
+	}
+
+	const rpcRequest = request as P2PRequest;
```
We can't cast the `request` object into a `P2PRequest ` because `P2PRequest` is a usable request object which can be responded to using its `request.end(...)` or `request.error(...)` method. Maybe this validation/sanitization function should just return a validated/sanitized `ProtocolRPCRequestPacket`; that's also more consistent with the `sanitizeProtocolMessage` function below. We will have to instantiate the `P2PRequest` as a separate step (not part of validation/sanitization); it needs the SocketCluster `respond` callback as third argument to its constructor.

shuse2 commented on packages/lisk-p2p/src/sanitization.ts:
```diff
@@ -92,3 +98,32 @@ export const sanitizePeerInfoList = (
+	return rpcRequest;
+};
+
+export const sanitizeProtocolMessage = (message: unknown): ProtocolMessagePacket => {
```
name of this should also be `validate`?

ishantiw replied:
actually before, all the functions were prefixed as process or validate. But later @jondubois renamed all of them to sanitize including the file name because these functions are not only validating but type casting or sanitizing it to some protocol type or preferred data structure.

shuse2 replied:
Mmm we have similar functionality in transaction with validating and the checking so maybe it’s better to standardize?

ishantiw replied:
Okay then I think it makes sense to just use `validate` here too  👍

shuse2 commented on packages/lisk-p2p/src/sanitization.ts:
```diff
@@ -14,8 +14,14 @@
  */
```
also, filename should be consistent, and all the function in this file?

ishantiw replied:
same here, should we rename it back to `validate` prefix? what do you think @jondubois ?

ishantiw committed:
:recycle: Refactor emit event in handlers and rename sanitize->validate

ishantiw committed:
:recycle: Use ProtocolRPCRequestPacket instead of P2PRequest and rename santization.ts -> validation.ts in test

ishantiw commented:
Addressed @jondubois @shuse2
